### DISCUSSION TITLE
Why I get CS0165 when the assignment is inside a try block?

### DISCUSSION THREAD

**Discussion by a510** [dotnet/roslyn#65920] [Q&A] [answered]
In the method below I get `CS0165: Use of unassigned local variable` although `x` is assigned and I can return it.
Why is this not allowed ?

``` csharp
int M()
{
    int x;
    try { x = 0; }
    finally { Console.WriteLine(x);  /* CS0165 */ }
    return x;
}
```

**Comment by CyrusNajmabadi** [comment, marked as answer]
Because an exception may be thrown before the assignment happens, thus causing x to be unassigned.

**Reply by a510** [reply]
You mean that just entering a `try` block might cause an exception?

**Reply by CyrusNajmabadi** [reply]
From the perspective of the language, an exception can occur at any time for any reason.

**Reply by svick** [reply]
> You mean that just entering a try block might cause an exception?

As far as the language is concerned, yes.

In practice, `Thread.Abort` on .Net Framework and `ControlledExecution.Run` on .Net 7 can cause an exception to be thrown pretty much at any point in the program, which is why the language assumes that an exception can be thrown anywhere.

**Reply by Flithor** [reply]
Interestingly, writing like this doesn't won't get CS0165.
Is it because the compiler will optimizes the if(true) statement?
```
int M()
{
    int x;
    if (true)
        x = 0;
    else
        throw new Exception();
    Console.WriteLine(x);
}
```

**Reply by CyrusNajmabadi** [reply]
Not sure what you're asking. In that code x will always be definitely assigned by the time the Console.Write is executed.

**Reply by svick** [reply]
@Flithor 

There are two things:

1. The compiler does understand that `if (true)` means that the `else` clause will never be executed. Because of this, the code will still compile if you change it to: `if (true) x = 0; else {}`.
2. The compiler understands that `throw` means that the `Console.WriteLine()` will not be executed. Because of this, this code also still compiles: `if (cond) x = 0; else throw new Exception();`.

[Both versions in SharpLab.](https://sharplab.io/#v2:C4LgbgpgTglgZgTwD4AEBMBGAsAKF+gAgGFcBvXAyglAFgIFkAKASgqvJyq4JgDtgCADwDcbbjzgFGwKAFcIrTuO6CCAXgIAGUUvEQANgGcIY5ZVIBfU1xQYAnI0HMdXK7oKnaDRgCMA9n76BADGfrwAJopcHGZ8AiLWVPBSoRFRZlSqGtqJlAbGudzAABZQfgDuBLwQlQCigsEQAA7AMGEsLsq2Dk6dlBZcuBZAA===)
